src/doublylinkedlist.c

```diff
@@ -110,8 +113,47 @@ void DList_InsertHeadList(PDLIST_ENTRY listHead, PDLIST_ENTRY entry)
 
 IMPLEMENT_MOCKABLE_FUNCTION(, int, DList_ForEach, PDLIST_ENTRY, listHead, DLIST_ACTION_FUNCTION, actionFunction, void*, actionContext)
 {
-    (void)listHead;
-    (void)actionFunction;
-    (void)actionContext;
-    return 0;
+    int result = 0;
+    /*Codes_SRS_DLIST_43_001: [If listHead is NULL, DList_ForEach shall fail and return a non - zero value.]*/
+    /*Codes_SRS_DLIST_43_002 : [If actionFunction is NULL, DList_ForEach shall fail and return a non - zero value.]*/
+    if (
+        listHead == NULL ||
+        actionFunction == NULL
+        )
+    {
+        /*Codes_SRS_DLIST_43_012: [If there are any failures, DList_ForEach shall fail and return a non - zero value.]*/
+        LogError("Invalid args: PDLIST_ENTRY listHead = %p, DLIST_ACTION_FUNCTION actionFunction = %p", listHead, actionFunction);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        PDLIST_ENTRY entry = listHead->Flink;
+        while (entry != listHead)
+        {
+            bool continueProcessing;
+            /*Codes_SRS_DLIST_43_009: [DList_ForEach shall call actionFunction on each entry in the list defined by listHead along with actionContext.]*/
+            if (actionFunction(entry, actionContext, &continueProcessing) != 0)
+            {
+                LogError("failure in actionFunction(entry, actionContext, &continueProcessing)");
```

Comment:
>entry, actionContext, &continueProcessing [](http://example.com/codeflow?start=52&length=41)

parameteres to actionFunction should be logged as values too.

There's never enough logging in the world of services.